Add support for building libdispatch on FreeBSD

Rogiel opened the pull request:
This mostly adds `defined(__FreeBSD__)` conditionals where necessary. I am 99% sure this will break builds on Darwin, specially on kevent stuff. I tried to build the Xcode project, but there are a LOT of files missing which probably are internal Apple headers and sources.

Please note that the library is not yet completely working on FreeBSD. It compiles, few tests are working, but most are still failing.

```
Test project /home/rogiel/Swift/swift-corelibs-libdispatch/build
      Start  1: dispatch_apply
 1/21 Test  #1: dispatch_apply ...................   Passed    0.15 sec
      Start  2: dispatch_api
 2/21 Test  #2: dispatch_api .....................***Failed    0.13 sec
      Start  3: dispatch_debug
 3/21 Test  #3: dispatch_debug ...................   Passed    0.12 sec
      Start  4: dispatch_queue_finalizer
 4/21 Test  #4: dispatch_queue_finalizer .........***Timeout  30.27 sec
      Start  5: dispatch_group
 5/21 Test  #5: dispatch_group ...................***Failed    0.13 sec
      Start  6: dispatch_overcommit
 6/21 Test  #6: dispatch_overcommit ..............***Timeout  30.01 sec
      Start  7: dispatch_context_for_key
 7/21 Test  #7: dispatch_context_for_key .........***Failed    0.13 sec
      Start  8: dispatch_after
 8/21 Test  #8: dispatch_after ...................***Failed    0.13 sec
      Start  9: dispatch_timer
 9/21 Test  #9: dispatch_timer ...................***Timeout  30.01 sec
      Start 10: dispatch_timer_short
10/21 Test #10: dispatch_timer_short .............***Timeout  30.01 sec
      Start 11: dispatch_timer_timeout
11/21 Test #11: dispatch_timer_timeout ...........***Failed    6.15 sec
      Start 12: dispatch_sema
12/21 Test #12: dispatch_sema ....................   Passed    0.12 sec
      Start 13: dispatch_timer_bit31
13/21 Test #13: dispatch_timer_bit31 .............***Timeout  30.01 sec
      Start 14: dispatch_timer_bit63
14/21 Test #14: dispatch_timer_bit63 .............***Timeout  30.02 sec
      Start 15: dispatch_timer_set_time
15/21 Test #15: dispatch_timer_set_time ..........***Timeout  30.02 sec
      Start 16: dispatch_starfish
16/21 Test #16: dispatch_starfish ................***Timeout  30.01 sec
      Start 17: dispatch_data
17/21 Test #17: dispatch_data ....................***Failed    0.13 sec
      Start 18: dispatch_io_net
18/21 Test #18: dispatch_io_net ..................***Failed   25.30 sec
      Start 19: dispatch_select
19/21 Test #19: dispatch_select ..................***Failed    0.12 sec
      Start 20: dispatch_c99
20/21 Test #20: dispatch_c99 .....................***Failed    0.13 sec
      Start 21: dispatch_plusplus
21/21 Test #21: dispatch_plusplus ................***Failed    0.13 sec
```

Rogiel commented:
For those now following swift-corelibs-dev@:
> kevent is very (very, very, very!) ugly right now and 99% sure broken on Darwin. FreeBSD and Darwin diverge on a ton of type definitions on kevent structures. I had to add a lot of casting to get it working on FreeBSD which almost certainly broken it on Darwin. Maybe it should be split into a kevent_mach for Apple platforms and another for FreeBSD, but that would cause code duplication problems.

alblue commented on CMakeLists.txt:
```diff
@@ -73,7 +75,8 @@ option(BUILD_SHARED_LIBS "build shared libraries" ON)
 if(CMAKE_SYSTEM_NAME STREQUAL Linux OR
-   CMAKE_SYSTEM_NAME STREQUAL Android)
+   CMAKE_SYSTEM_NAME STREQUAL Android OR
+   CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
```
This would be a less intrusive patch if you put the conditional above Android rather than below. If you had it above then you'd only generate one line of addition:

```
 if(CMAKE_SYSTEM STREQUAL Linux OR
+   CMAKE_SYSTEM STREQUAL FreeBSD OR
   CMAKE_SYSTEM STREQUAL Android)
```

alblue commented on src/CMakeLists.txt:
```diff
@@ -123,11 +135,13 @@ target_include_directories(dispatch
                              ${CMAKE_CURRENT_SOURCE_DIR}
                              ${CMAKE_CURRENT_BINARY_DIR}
                              ${CMAKE_SOURCE_DIR}/private)
+
```
I'd recommend against adding blank lines as it adds more noise to the patch than is necessary to get your changes through

alblue commented on src/block.cpp:
```diff
@@ -28,6 +28,8 @@
 #error Must build without C++ exceptions
 #endif
 
+#define _Bool bool
```
This looks like it might cause problems for other builds

Rogiel replied:
Indeed. This shouldn't even be there!

alblue commented on src/event/event_config.h:
```diff
@@ -76,16 +76,19 @@
 #		define DISPATCH_USE_KEVENT_QOS 0
 #	endif
 
+/// The flags here are set unconditionally (from DISPATCH_USE_KEVENT_QOS).
```
This also looks like an odd change - why has the indentation changed?

Rogiel replied:
These flags were being used everywhere and unconditionally. Originally they were only defined if `defined(EV_SET_QOS)`, which is not true on FreeBSD. The indentation didn't change per se, it was moved one level down and indentation corrected accordingly.

 - In the epoll `KEVENT_FLAG_IMMEDIATE` is emulated with a timeout, the same must be done on kevent (this is not done yet -- this PR is only to get the library building and submitting things in small batches)
 - `KEVENT_FLAG_ERROR_EVENTS` already has a emulation (line 611)

alblue commented on src/event/event_kevent.c:
```diff
@@ -1129,7 +1152,7 @@ _dispatch_event_loop_poke(dispatch_wlh_t wlh, uint64_t dq_state, uint32_t flags)
 			.filter = EVFILT_USER,
 			.fflags = NOTE_TRIGGER,
-			.udata = (uintptr_t)DISPATCH_WLH_MANAGER,
+			.udata = (dispatch_kqueue_udata)DISPATCH_WLH_MANAGER,
```
There seems to be a lot of these kind of changes here. It might make sense to break this out (`uintptr_t` -> `dispatch_kqueue_udata`) into its own change first, which can then be merged ahead of the rest of the changes.

Rogiel replied:
Yeah... I think kevent is a horrible mess with these changes, I hate it, but I don't see it could be made much better than this.

I think it would make sense to split into separate changes.

alblue commented on src/event/event_kevent.c:
```diff
@@ -1271,6 +1294,8 @@ _dispatch_event_loop_timer_program(uint32_t tidx,
 #endif
 	};
 
+	(void) leeway; // if DISPATCH_HAVE_TIMER_COALESCING == 0
```
This change seems odd ...

Rogiel replied:
This is for `-Werror` compliance. Maybe it should be:

```
DISPATCH_NOINLINE
static void
_dispatch_event_loop_timer_program(uint32_t tidx,
		uint64_t target, uint64_t leeway, uint16_t action)
{
	dispatch_kevent_s ke = {
		.ident = DISPATCH_KEVENT_TIMEOUT_IDENT_MASK | tidx,
		.filter = EVFILT_TIMER,
		.flags = action | EV_ONESHOT,
		.fflags = _dispatch_timer_index_to_fflags[tidx],
		.data = (int64_t)target,
		.udata = (dispatch_kqueue_udata)&_dispatch_timers_heap[tidx],
#if DISPATCH_HAVE_TIMER_COALESCING
		.ext[1] = leeway,
#endif
#if DISPATCH_USE_KEVENT_QOS
		.qos = _PTHREAD_PRIORITY_EVENT_MANAGER_FLAG,
#endif
	};

#if !DISPATCH_HAVE_TIMER_COALESCING
	(void) leeway;
#endif

	_dispatch_kq_deferred_update(DISPATCH_WLH_ANON, &ke);
}
```

(added the whole function for more context)

MadCoder replied:
no the change isn't odd, this is because leeway is unused, and it's fine to do it in an unguarded part. `(void)foo;` to avoid an unused warning is idiomatic in libdispatch.

Rogiel commented on src/event/event_kevent.c:
```diff
@@ -601,7 +614,7 @@ _dispatch_kq_poll(dispatch_wlh_t wlh, dispatch_kevent_t ke, int n,
 			ke[r].flags |= EV_RECEIPT;
 		}
-		out_n = n;
+		n_out = n;
```
Bug in the existing code. This is probably due to a refactoring and the fact that code with `!DISPATCH_USE_KEVENT_QOS` not being exercised recently.

Rogiel commented on src/CMakeLists.txt:
```diff
@@ -100,7 +111,8 @@ if(ENABLE_SWIFT)
                       -I ${CMAKE_SOURCE_DIR}
-                      ${swift_optimization_flags})
+                      -I/usr/include
+          ${swift_optimization_flags})
```
I believe this is a issue with swiftc driver issue, for some reason it appears to not be appending /usr/include to path. Maybe I should make it a variable and mark it explicitly as a workaround.

I tried to find the root issue in the swift compiler but I couldn't.

And yes, indentation has changed... my IDE insists in reformatting after each line and messed on this and some other places too.

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -337,7 +350,7 @@ DISPATCH_ALWAYS_INLINE
 _dispatch_kevent_get_unote(dispatch_kevent_t ke)
 {
-	dispatch_assert((ke->udata & DISPATCH_KEVENT_MUXED_MARKER) == 0);
+	dispatch_assert(((uintptr_t)ke->udata & DISPATCH_KEVENT_MUXED_MARKER) == 0);
```
please add a helper just above this get_unote to hide this ugly looking cast:

```C
DISPATCH_ALWAYS_INLINE
static inline bool
_dispatch_kevent_unote_is_muxed(dispatch_kevent_t ke)
{
    return (uintptr_t)ke->udata & DISPATCH_KEVENT_MUXED_MARKER;
}
```

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -356,9 +369,9 @@ _dispatch_kevent_print_error(dispatch_kevent_t ke)
 		// for EV_DELETE if the update was deferred we may have reclaimed
 		// the udata already, and it is unsafe to dereference it now.
-	} else if (ke->udata & DISPATCH_KEVENT_MUXED_MARKER) {
+	} else if ((uintptr_t)ke->udata & DISPATCH_KEVENT_MUXED_MARKER) {
```
use the inline

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -356,9 +369,9 @@ _dispatch_kevent_print_error(dispatch_kevent_t ke)
+	} else if ((uintptr_t)ke->udata & DISPATCH_KEVENT_MUXED_MARKER) {
 		ke->flags |= _dispatch_kevent_get_muxnote(ke)->dmn_kev.flags;
-	} else if (ke->udata) {
+	} else if ((uintptr_t)ke->udata) {
```
not needed

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -463,7 +476,7 @@ _dispatch_kevent_drain(dispatch_kevent_t ke)
 #endif
 
-	if (ke->udata & DISPATCH_KEVENT_MUXED_MARKER) {
+	if ((uintptr_t)ke->udata & DISPATCH_KEVENT_MUXED_MARKER) {
```
use the inline

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -32,6 +32,14 @@
 #define DISPATCH_MACH_AUDIT_TOKEN_PID (5)
 
+#if defined(__FreeBSD__)
+typedef void* dispatch_kqueue_udata;
```
The name is wrong, udata is a kevent field, not udata. `dispatch_kevent_udata_t` would make more sense, and define it once this way:


```C
#define dispatch_kevent_udata_t  typeof(((dispatch_kevent_t)NULL)->udata)
```

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -480,7 +493,7 @@ _dispatch_kq_create(const void *guard_ptr)
 		.filter = EVFILT_USER,
 		.flags = EV_ADD|EV_CLEAR,
-		.udata = (uintptr_t)DISPATCH_WLH_MANAGER,
+		.udata = (dispatch_kqueue_udata)DISPATCH_WLH_MANAGER,
```
`(dispatch_kevent_udata_t)`

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -542,7 +555,7 @@ _dispatch_kq_init(void *context)
 			.flags = EV_ADD|EV_CLEAR,
 			.qos = _PTHREAD_PRIORITY_EVENT_MANAGER_FLAG,
-			.udata = (uintptr_t)DISPATCH_WLH_MANAGER,
+			.udata = (dispatch_kqueue_udata)DISPATCH_WLH_MANAGER,
```
`(dispatch_kevent_udata_t)`

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -601,7 +614,7 @@ _dispatch_kq_poll(dispatch_wlh_t wlh, dispatch_kevent_t ke, int n,
 			ke[r].flags |= EV_RECEIPT;
 		}
-		out_n = n;
+		n_out = n;
```
🤦‍♂️ @ myself

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -614,6 +627,8 @@ _dispatch_kq_poll(dispatch_wlh_t wlh, dispatch_kevent_t ke, int n,
 		}
 		r = kevent_qos(kqfd, ke, n, ke_out, n_out, buf, avail, flags);
 #else
+		(void) buf; // buf is unused here
```
remove the spaces

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -717,19 +732,20 @@ _dispatch_kq_unote_set_kevent(dispatch_unote_t _du, dispatch_kevent_t dk,
 		.filter = dst->dst_filter,
 		.flags  = flags,
-		.udata  = (uintptr_t)du,
+		.udata  = (dispatch_kqueue_udata)du,
```
`(dispatch_kevent_udata_t)`

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -717,19 +732,20 @@ _dispatch_kq_unote_set_kevent(dispatch_unote_t _du, dispatch_kevent_t dk,
 		.qos    = (typeof(dk->qos))pp,
 #endif
 	};
+	(void) pp; // if DISPATCH_USE_KEVENT_QOS == 0
```
remove the space and the comment

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -717,19 +732,20 @@ _dispatch_kq_unote_set_kevent(dispatch_unote_t _du, dispatch_kevent_t dk,
 static inline int
 _dispatch_kq_deferred_find_slot(dispatch_deferred_items_t ddi,
-		int16_t filter, uint64_t ident, uint64_t udata)
+		int16_t filter, uint64_t ident, dispatch_kqueue_udata udata)
```
`dispatch_kevent_udata_t`

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -825,7 +841,7 @@ _dispatch_kq_unote_update(dispatch_wlh_t wlh, dispatch_unote_t _du,
 	if (ddi && wlh == _dispatch_get_wlh()) {
 		int slot = _dispatch_kq_deferred_find_slot(ddi,
-				du->du_filter, du->du_ident, (uintptr_t)du);
+				du->du_filter, du->du_ident, (dispatch_kqueue_udata)du);
```
`(dispatch_kevent_udata_t)` wrap to 80 columns

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -961,7 +979,7 @@ _dispatch_unote_register_muxed(dispatch_unote_t du, dispatch_wlh_t wlh)
 		dmn->dmn_kev.qos = _PTHREAD_PRIORITY_EVENT_MANAGER_FLAG;
 #endif
-		dmn->dmn_kev.udata = (uintptr_t)dmn | DISPATCH_KEVENT_MUXED_MARKER;
+		dmn->dmn_kev.udata = (dispatch_kqueue_udata) ((uintptr_t)dmn | DISPATCH_KEVENT_MUXED_MARKER);
```
remove the space, wrap to 80 columns

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -1129,7 +1152,7 @@ _dispatch_event_loop_poke(dispatch_wlh_t wlh, uint64_t dq_state, uint32_t flags)
 			.filter = EVFILT_USER,
 			.fflags = NOTE_TRIGGER,
-			.udata = (uintptr_t)DISPATCH_WLH_MANAGER,
+			.udata = (dispatch_kqueue_udata)DISPATCH_WLH_MANAGER,
```
`(dispatch_kevent_udata_t)`

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -1262,7 +1285,7 @@ _dispatch_event_loop_timer_program(uint32_t tidx,
 		.fflags = _dispatch_timer_index_to_fflags[tidx],
 		.data = (int64_t)target,
-		.udata = (uintptr_t)&_dispatch_timers_heap[tidx],
+		.udata = (dispatch_kqueue_udata)&_dispatch_timers_heap[tidx],
```
`(dispatch_kevent_udata_t)`

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -1271,6 +1294,8 @@ _dispatch_event_loop_timer_program(uint32_t tidx,
 #endif
 	};
 
+	(void) leeway; // if DISPATCH_HAVE_TIMER_COALESCING == 0
```
remove space, empty line and comment.

MadCoder commented on src/event/workqueue.c:
```diff
@@ -91,6 +93,9 @@ _dispatch_workq_worker_register(dispatch_queue_t root_q, qos_class_t cls)
 	mon->registered_tids[worker_id] = tid;
 	_dispatch_unfair_lock_unlock(&mon->registered_tid_lock);
+#else
+	(void) root_q;
```
remove spaces

MadCoder commented on src/event/workqueue.c:
```diff
@@ -113,6 +118,9 @@ _dispatch_workq_worker_unregister(dispatch_queue_t root_q, qos_class_t cls)
 	}
 	_dispatch_unfair_lock_unlock(&mon->registered_tid_lock);
+#else
+	(void) root_q;
```
remove spaces

MadCoder commented on src/init.c:
```diff
@@ -737,7 +737,7 @@ _dispatch_logv_file(const char *msg, va_list ap)
 #if DISPATCH_DEBUG
 	offset += dsnprintf(&buf[offset], bufsiz - offset, "%llu\t",
-			_dispatch_absolute_time() - dispatch_log_basetime);
+			(unsigned long long) _dispatch_absolute_time() - dispatch_log_basetime);
```
remove the space after the cast, wrap to 80 columns if needed

MadCoder commented on src/init.c:
```diff
@@ -834,7 +834,7 @@ _dispatch_debugv(dispatch_object_t dou, const char *msg, va_list ap)
 #if DISPATCH_DEBUG && !DISPATCH_USE_OS_DEBUG_LOG
 	offset += dsnprintf(&buf[offset], bufsiz - offset, "%llu\t\t%p\t",
-			_dispatch_absolute_time() - dispatch_log_basetime,
+			(unsigned long long) _dispatch_absolute_time() - dispatch_log_basetime,
```
remove the space after the cast, wrap to 80 columns if needed

MadCoder commented on src/io.c:
```diff
@@ -419,7 +424,7 @@ dispatch_io_create_with_path(dispatch_io_type_t type, const char *path,
 		_dispatch_io_syscall_switch_noerr(err,
 			(path_data->oflag & O_NOFOLLOW) == O_NOFOLLOW
-#ifndef __linux__
+#if !defined(__linux__) && !defined(__FreeBSD__)
```
`#if __APPLE`

MadCoder commented on src/io.c:
```diff
@@ -627,7 +632,7 @@ dispatch_io_set_interval(dispatch_io_t channel, uint64_t interval,
 	_dispatch_retain(channel);
 	dispatch_async(channel->queue, ^{
-		_dispatch_channel_debug("set interval: %llu", channel, interval);
+		_dispatch_channel_debug("set interval: %llu", channel, (unsigned long long) interval);
```
remove the space after the cast, wrap to 80 columns

MadCoder commented on src/io.c:
```diff
@@ -1372,7 +1377,7 @@ _dispatch_fd_entry_create_with_fd(dispatch_fd_t fd, uintptr_t hash)
 				);
 			}
-			dev_t dev = major(st.st_dev);
+			dev_t dev = (dev_t) major(st.st_dev);
```
remove the space after the cast

MadCoder commented on src/io.c:
```diff
@@ -1459,7 +1464,7 @@ _dispatch_fd_entry_create_with_path(dispatch_io_path_data_t path_data,
 	_dispatch_fd_entry_debug("create: path %s", fd_entry, path_data->path);
 	if (S_ISREG(mode)) {
-		_dispatch_disk_init(fd_entry, major(dev));
+		_dispatch_disk_init(fd_entry, (dev_t) major(dev));
```
ditto

MadCoder commented on src/queue.c:
```diff
@@ -914,6 +914,13 @@ gettid(void)
+static inline pid_t
+gettid(void)
+{
+	return (pid_t) pthread_getthreadid_np();
```
remove the space.

MadCoder commented on src/queue.c:
```diff
@@ -4460,7 +4467,7 @@ _dispatch_runloop_queue_class_poke(dispatch_queue_t dq)
 	}
 
-#if HAVE_MACH
+#if defined(HAVE_MACH)
```
revert, HAVE_MACH is defined to 0 sometimes

MadCoder commented on src/queue.c:
```diff
@@ -6138,6 +6149,29 @@ _dispatch_runloop_queue_handle_init(void *ctxt)
 		}
 	}
 	handle = fd;
+#elif defined(__FreeBSD__)
```
looks like the whole switch is the same, it should be shared better

MadCoder commented on src/queue.c:
```diff
@@ -6219,13 +6253,13 @@ void
 {
 	_dispatch_root_queues_init();
-#if HAVE_PTHREAD_MAIN_NP
+#if defined(HAVE_PTHREAD_MAIN_NP)
```
no, it is set to 0 not undef when not there, revert

MadCoder commented on src/queue.c:
```diff
@@ -6239,7 +6273,7 @@ dispatch_main(void)
 		pthread_exit(NULL);
 		DISPATCH_INTERNAL_CRASH(errno, "pthread_exit() returned");
-#if HAVE_PTHREAD_MAIN_NP
+#if defined(HAVE_PTHREAD_MAIN_NP)
```
revert

MadCoder commented on src/queue.c:
```diff
@@ -6219,13 +6253,13 @@ void
 		_dispatch_program_is_probably_callback_driven = true;
 		_dispatch_ktrace0(ARIADNE_ENTER_DISPATCH_MAIN_CODE);
-#ifdef __linux__
+#if defined(__linux__)
```
avoid gratuitous changes that don't change semantics, ifdef is fine, revert

MadCoder commented on src/shims.h:
```diff
@@ -50,11 +52,11 @@
 #endif // HAVE_PTHREAD_WORKQUEUES
 
-#if DISPATCH_USE_INTERNAL_WORKQUEUE
+#ifdef DISPATCH_USE_INTERNAL_WORKQUEUE
```
no, broken, revert. this thing is defined to 0 sometimes

MadCoder commented on src/shims.h:
```diff
@@ -50,11 +52,11 @@
 #endif
 
-#if HAVE_PTHREAD_NP_H
+#ifdef HAVE_PTHREAD_NP_H
```
revert

Rogiel replied:
```
error: expected value in expression
#if HAVE_PTHREAD_NP_H
```

This should be defined by CMake, correct?

MadCoder replied:
yes, also libdispatch doesn't build cleanly with `-Wundef` today, which we should likely fix long term, but moving random `#if` to `#ifdef` will break things.

Rogiel replied:
I looked further into this, and it seems ifdef is correct.

config.h.in uses #cmakedefine which does not evaluate to 1 or 0 but to defined or not defined, as stated in CMake documentation: https://cmake.org/cmake/help/latest/command/configure_file.html

MadCoder replied:
then the fix is to make it `cmakedefine01`

Darwin doesn't use CMake and we do define some of these HAVE_* things to 0 internally not ifdef/ifndef.

MadCoder commented on src/shims/atomic.h:
```diff
@@ -31,6 +31,7 @@
 #error libdispatch requires C11 with <stdatomic.h>
 #endif
 
+#include <stdbool.h> // FreeBSD is requiring it, wont hurt on other platforms
```
lose the comment

MadCoder commented on src/shims/freebsd_stubs.h:
```diff
@@ -0,0 +1,66 @@
+/*
```
feels very similar to linux, we should share more of these

MadCoder commented on src/shims/lock.c:
```diff
@@ -338,6 +338,20 @@ _dispatch_ulock_wake(uint32_t *uaddr, uint32_t flags)
+#else
+
+static int
+_dispatch_ulock_wait(uint32_t *uaddr, uint32_t val, uint32_t timeout,
```
please do not do that, making the build pass with such a critical piece missing is not helping. please revert.

MadCoder commented on src/shims/lock.c:
```diff
@@ -366,6 +380,21 @@ _dispatch_unfair_lock_wake(uint32_t *uaddr, uint32_t flags)
 
+#elif HAVE_UMTX
+
+// returns 0, ETIMEDOUT, ENOTEMPTY, EFAULT
```
same

MadCoder commented on src/shims/lock.c:
```diff
@@ -540,6 +569,15 @@ _dispatch_unfair_lock_lock_slow(dispatch_unfair_lock_t dul,
 	(void)flags;
 	_dispatch_futex_lock_pi(&dul->dul_lock, NULL, 1, FUTEX_PRIVATE_FLAG);
 }
+#elif HAVE_UMTX
```
same

MadCoder commented on src/shims/lock.h:
```diff
@@ -79,6 +79,28 @@ _dispatch_lock_owner(dispatch_lock lock_value)
 	return lock_value & DLOCK_OWNER_MASK;
 }
 
+#elif defined(__FreeBSD__)
```
this is absolutely not functional, please do not add this until it is properly ported.

MadCoder commented on src/swift/Source.swift:
```diff
@@ -112,7 +112,7 @@ public extension DispatchSource {
 #endif
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
FreeBSD has EVFILT_PROC, this is wrong

MadCoder commented on src/swift/Source.swift:
```diff
@@ -170,7 +170,7 @@ public extension DispatchSource {
 #endif
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
I'm fairly sure FreeBSD has these, feels wrong to me.

MadCoder commented on src/swift/Source.swift:
```diff
@@ -207,7 +207,7 @@ public extension DispatchSource {
 	}
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
FreeBSD has EVFILT_VNODE, this is wrong

MadCoder commented on src/swift/Source.swift:
```diff
@@ -260,7 +260,7 @@ public extension DispatchSourceMemoryPressure {
 #endif
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
FreeBSD has EVFILT_PROC, this is wrong

MadCoder commented on src/swift/Source.swift:
```diff
@@ -618,7 +618,7 @@ public extension DispatchSourceTimer {
 }
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
FreeBSD has EVFILT_VNODE, this is wrong

MadCoder commented on src/swift/Source.swift:
```diff
@@ -690,7 +690,7 @@ internal func _swift_dispatch_source_type_mach_recv() -> dispatch_source_type_t
 #endif
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
FreeBSD has EVFILT_PROC, this is wrong

MadCoder commented on src/swift/Source.swift:
```diff
@@ -704,7 +704,7 @@ internal func _swift_dispatch_source_type_signal() -> dispatch_source_type_t
 internal func _swift_dispatch_source_type_timer() -> dispatch_source_type_t
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
FreeBSD has EVFILT_VNODE, this is wrong

MadCoder commented on src/swift/Wrapper.swift:
```diff
@@ -180,7 +180,7 @@ extension DispatchSource : DispatchSourceMachSend,
 #endif
 
-#if !os(Linux) && !os(Android)
+#if !os(Linux) && !os(Android) && !os(FreeBSD)
```
revert all in this, FreeBSD has all the underlying filters

MadCoder commented on src/transform.c:
```diff
@@ -22,14 +22,22 @@
 #ifdef __APPLE__
 #include <libkern/OSByteOrder.h>
-#elif __linux__
+#elif defined(__linux__)
```
gratuitous change, please revert

MadCoder commented on tests/bsdtests.h:
```diff
@@ -29,6 +29,10 @@
 #endif
 
+#if defined(__linux__) || defined(__FreeBSD__)
+#include <unistd.h>
```
just always `#include it` it will not harm, and remove the empty lines

MadCoder commented on tests/dispatch_test.c:
```diff
@@ -56,7 +56,11 @@ dispatch_test_check_evfilt_read_for_fd(int fd)
+#if defined(__APPLE__)
 		.ident = fd,
+#else
+		.ident = (uintptr_t) fd,
```
always do the cast it will not harm

MadCoder commented on tests/freebsd_port.h:
```diff
@@ -0,0 +1,54 @@
+#include <limits.h>
```
seems a lot is like the linux port header, same as other comments, should be more of a "non_darwin_port.h"

Rogiel commented on private/queue_private.h:
```diff
@@ -227,7 +227,7 @@ DISPATCH_EXPORT DISPATCH_MALLOC DISPATCH_RETURNS_RETAINED DISPATCH_WARN_RESULT
 dispatch_queue_t
 dispatch_pthread_root_queue_create(const char *_Nullable label,
-	unsigned long flags, const pthread_attr_t *_Nullable attr,
+	unsigned long flags, _Nullable const pthread_attr_t *_Nullable attr,
```
This must be a Clang bug then?

/home/rogiel/Swift/swift-corelibs-libdispatch/private/queue_private.h:230:29: error: pointer is missing a nullability type specifier
      (_Nonnull, _Nullable, or _Null_unspecified) [-Werror,-Wnullability-completeness]

MadCoder replied:
maybe not, if pthread_attr_t is a pointer type on FreeBSD, but it's not on Darwin.

Rogiel replied:
Yeah... thats absolutely the case then.

Rogiel replied:
Not sure where I should and and #ifdef for this... the entire function signature? only that line? define a `#define NULLBLE_PTHREAD_ATTR _Nullable` on FreeBSD and `#define NULLBLE_PTHREAD_ATTR` on other platforms?

dgrove-oss commented on os/freebsd_base.h:
```diff
@@ -0,0 +1,125 @@
+/*
```
I agree; a generic_unix_base with a few small #ifdef clauses will be easier to maintain.

Rogiel commented:
What I have done:
 - Reverted all swift changes: those weren't supposed to be there anyway and had no testing whatsoever, just some find and replace.
 - Reverted lock stubs
 - Reverted most changes to queue

What still remains to be done:
 - [ ] Fix to nullability in `attr` argument of `dispatch_pthread_root_queue_create` (`pthread_attr_t` is a pointer on FreeBSD but not on Darwin)

Rogiel commented:
All right @MadCoder, I have fixed the `pthread_attr_t` issue with a conditionally defined macro. I find this solution pretty ugly, but I don't know what do to here.

I guess the ugly macro is better than redefining the whole signature? It a private header anyways, maybe it's not that bad.

Just to recap, these changes makes the library *almost* compile on FreeBSD. There are still 4 files with issues:
 - queue.c (mostly runloop implementation)
 - lock.c/lock.h shims
 - bsdtestharness.c (this is a new issue, wasn't there in the first patch)

MadCoder commented on os/generic_unix_base.h:
```diff
@@ -10,11 +10,15 @@
 
 #if __has_include(<sys/sysmacros.h>)
 #include <sys/sysmacros.h>
+#if defined(__FreeBSD)
```
shouldn't this be `__FreeBSD__` ?

MadCoder replied:
also it's weird it's inside the `__has_include(<sys/sysmacros.h>)`, I'd move it outside.

MadCoder commented on src/event/event_config.h:
```diff
@@ -76,16 +76,14 @@
 #	endif
 
+#	ifndef KEVENT_FLAG_ERROR_EVENTS
+#	 define KEVENT_FLAG_ERROR_EVENTS 0x002
```
the indent is wrong here

MadCoder commented on src/event/event_config.h:
```diff
@@ -106,6 +104,14 @@
 #	define NOTE_FUNLOCK 0x00000100
 #	endif
 
+/// FreeBSD's kevent does not support those
```
double `//` triple is for headerdoc.

MadCoder commented on src/event/event_config.h:
```diff
@@ -146,9 +152,15 @@
+// translated back by the various backends to similar semantics
+// hence must be defined even on non Darwin platforms
+#ifndef KEVENT_FLAG_IMMEDIATE
+# define KEVENT_FLAG_IMMEDIATE 0x001
```
remove the leading space

Rogiel replied:
All other #defines are with the leading space on this file... should I still remove them?

MadCoder replied:
the big hunk of configuration has a lot of nesting so it is pseudo indented for readability, but once it's at top level we don't use leading indent.

ALso I think it is tab-indented for the big fat one at the top not space ;)

Rogiel replied:
@MadCoder changed to a tab AND fixed my tab size... now I can see them 😝 

I kept a tab for consistency with the previous defines.

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -614,6 +631,8 @@ _dispatch_kq_poll(dispatch_wlh_t wlh, dispatch_kevent_t ke, int n,
 		}
 		r = kevent_qos(kqfd, ke, n, ke_out, n_out, buf, avail, flags);
 #else
+		(void)buf; // buf is unused here
```
remove the comment, `(void)variable` means it's unused throughout the repository already

MadCoder commented on src/event/event_kevent.c:
```diff
@@ -961,7 +983,8 @@ _dispatch_unote_register_muxed(dispatch_unote_t du, dispatch_wlh_t wlh)
 #endif
-		dmn->dmn_kev.udata = (uintptr_t)dmn | DISPATCH_KEVENT_MUXED_MARKER;
+		dmn->dmn_kev.udata = (dispatch_kevent_udata_t)((uintptr_t)dmn |
+		  DISPATCH_KEVENT_MUXED_MARKER);
```
this indent looks wrong, it should be led by 4 tabs (the 2 from the line above + 2 because it's a split line)

MadCoder commented on src/io.c:
```diff
@@ -2165,7 +2171,7 @@ _dispatch_operation_advise(dispatch_operation_t op, size_t chunk_size)
 	advise.ra_offset = op->advise_offset;
 	op->advise_offset += advise.ra_count;
-#ifdef __linux__
+#if __linux__
```
I think this change is wrong, `#if defined()` is what you use elswhere, please revert

MadCoder commented on src/shims/atomic.h:
```diff
@@ -31,6 +31,10 @@
 
+// FreeBSD only defines _Bool in C mode. In C++ mode _Bool is not being defined.
+#if defined(__cplusplus) && defined(__FreeBSD__)
+# define _Bool bool
```
please remove the leading space.

MadCoder commented on tests/dispatch_test.c:
```diff
@@ -56,7 +56,7 @@ dispatch_test_check_evfilt_read_for_fd(int fd)
 	assert(kq != -1);
 	struct kevent ke = {
-		.ident = fd,
+		.ident = (uintptr_t) fd,
```
remove space after cast

MadCoder commented:
Btw @Rogiel it seems your patch has been stuck for a long time if I read history right. For some reason, when someone pushes a commit on a branch I'm reviewing I get no notification. Do not hesitate to ping me explicitly in the comments with an @MadCoder, that definitely sends me an email that I react to (which is how I looked at it again).

I'm really sorry it felt like I forgot you for months, I just don't poll pull requests for lack of time.

MadCoder commented:
@swift-ci please test

Rogiel commented:
@MadCoder 

No problem, I just got busy too and my FreeBSD machine died 😕

Okay, I made the changes (except the one about the leading space in the a #define, see the inline comment). I will wait until CI is done in case anything else comes up.

MadCoder commented:
@swift-ci please test

MadCoder commented:
@swift-ci please test
